shell: Add SHELL_ALL-option to default enable all subsystem shells

ClaCodes opened the pull request:
Option to enable all subsystem shells by default. This option can be enabled during early app developement or on some testing builds. Allows for explorative familiarization with the available shell commands using shell-help and tab-completion to guide the exploration.

As zephyr is a big project, it can be overwelming for new users. Allow to set a single shell-option so that the respective shell for all subsystems are always available, whenever those subsystems are activated. This way users can explore the available commands.

- I am not so familiar with Kconfig. There might be a problem with the current proposal
- Alternative naming `SHELL_MAXIMAL`. Other?
- Left all subsystem shells that are already `default y if SHELL` or `default y if !SHELL_MINIMAL` unchanged

ClaCodes committed:
shell: Add SHELL_ALL-otion to default enable all subsystem shells

Option to enable all subsystem shells by default. This option can be
enabled during early app developement or on some testing builds. Allows for
explorative familiarization with the available shell commands using
shell-help and tab-completion.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
drivers: modem: shell: Do not select CONFIG_SHELL

Instead depend on it.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
net: shell: Do not select CONFIG_SHELL

Instead depend on it.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
net: lwm2m: shell: Do not select CONFIG_SHELL

Instead depend on it.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
net: l2: ieee802154: shell: Do not select CONFIG_SHELL

Instead depend on it.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
ztest: shell: Do not select CONFIG_SHELL

Instead depend on it.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
bluetooth: shell: Do not select CONFIG_SHELL

Instead depend on it.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
bluetooth: mesh: shell: Do not select CONFIG_SHELL

Instead depend on it.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
shell: Add missing shell dependency to various subsystem shells

To unify the experience, all subsystem shells should depend on the global
shell option.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

pdgendt commented on subsys/shell/Kconfig:
```diff
@@ -25,6 +25,13 @@ config SHELL_MINIMAL
 
+config SHELL_ALL
+	bool "Single config enable for all subsytem shells"
+	default n
```
Omit `default n`

pdgendt commented on subsys/shell/Kconfig:
```diff
@@ -25,6 +25,13 @@ config SHELL_MINIMAL
+	default n
+	depends on !SHELL_MINIMAL
+	help
+	  Enable all subsystem shells by default.
```
"Subsystem" isn't clear IMO, maybe something like "Default enable all available shell commands"?

kartben replied:
I think the commonly used terminology is shell modules

ClaCodes replied:
changed to `shell modules`

pdgendt commented on subsys/testsuite/ztest/Kconfig:
```diff
@@ -25,7 +25,8 @@ config ZTEST_TEST_DELAY_MS
 config ZTEST_SHELL
 	bool "Ztest with shell support"
-	select SHELL
+	depends on SHELL
+	default y if SHELL_ALL
```
I am in favour of this change, but make sure that `prj.conf` files that only had `CONFIG_ZTEST_SHELL` enabled now also have `CONFIG_SHELL`.

ClaCodes replied:
I could only find one. There I had already added the option. Are there more?

jukkar commented on subsys/net/l2/Kconfig:
```diff
@@ -69,6 +69,8 @@ endif # NET_L2_WIFI_MGMT
 
 config NET_L2_WIFI_SHELL
 	bool "Wi-Fi shell module"
+	depends on SHELL
```
What about depending here to NET_SHELL so user can disable network shell support by that symbol.
Same comment to other net related depends

jakub-uC commented:
I like the idea very much as well. My only concern, however, is that it might be difficult to maintain. How can we ensure that all newly introduced commands or subsystems implementing the shell will consistently include support for `SHELL_ALL` in `Kconfig`? At a minimum, I believe it would be advisable to explicitly highlight the requirement to support `CONFIG_SHELL_ALL` in the shell documentation: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/services/shell/index.rst

pdgendt commented:
I think the order of the commits currently breaks bisect-ability, with the introduction of `SHELL_ALL` you have created circular dependencies, and fixed those with later commits. It should be the other way around.

cfriedt commented on lib/posix/shell/Kconfig:
```diff
@@ -3,16 +3,14 @@
 
 menu "POSIX Shell Utilities"
 
-if SHELL
```
I think it would make sense to keep this

cfriedt commented on lib/posix/shell/Kconfig:
```diff
@@ -3,16 +3,14 @@
-
 config POSIX_SHELL
 	bool
+	depends on SHELL
```
```suggestion
```

cfriedt commented on lib/posix/shell/Kconfig:
```diff
@@ -3,16 +3,14 @@
 rsource "Kconfig.env"
 rsource "Kconfig.uname"
 
-endif # SHELL
-
```
I think we would like to keep this

cfriedt commented on lib/posix/shell/Kconfig.env:
```diff
@@ -3,6 +3,8 @@
 
 config POSIX_ENV_SHELL
 	bool "Support for shell"
+	depends on SHELL
```
```suggestion
```

cfriedt commented on lib/posix/shell/Kconfig.uname:
```diff
@@ -5,6 +5,8 @@ if POSIX_SINGLE_PROCESS
 
 config POSIX_UNAME_SHELL
 	bool "Support for `uname` command"
+	depends on SHELL
```
```suggestion
```

cfriedt reviewed (commented):
Just some minor nits for posix shell commands.

ClaCodes reviewed (commented):
- @jakub-uC agree. I think the best is to have a common pattern all over the code base and include it in the documentation (will do soon).
- @pdgendt reordered commit history
- @cfriedt reverted these commits. I still think it would be better the other way to help unify across the code base for maintainability reasons. Reverted to ease adoption and delay that discussion to separate PR.


I will add docs, adjust description and remove RFC label soon.

cfriedt commented on lib/posix/shell/Kconfig:
```diff
@@ -7,6 +7,7 @@ if SHELL
 
 config POSIX_SHELL
 	bool
+	default y if SHELL_ALL
```
This should be removed as well.

Note: this is a non-user-configurable (i.e. it is a common component selected by other Kconfig options).

ClaCodes committed:
doc: shell: Add section about `SHELL_ALL` option

Add a short section to the docs describing the use of the added `SHELL_ALL`
option.

ClaCodes commented:
- docs: added note about binary size

henrikbrixandersen reviewed (changes requested):
I am blocking this for now until further discussion has taken place as at least two maintainers have raised concerns about the value of this addition.

ClaCodes commented:
> Who is supposed to use this?

@henrikbrixandersen Application developers in early app developement. For example when porting to a new board to test all available subsystems. Furthermore builds where size is not so important and someone wants to explore all the possiblities available on their board without searching for every `CONFIG_.*SHELL` option. Furthermore testing build, that use some kind of systemtest frame work to test a firmware on the target board via the shell.

> Is it really worth maintaining a Kconfig option like this instead of just having users enable the shell modules, they actually need?

This has essentially to be decided by the community. Personally I think yes, so I proposed it. So far feedback was rather positive so I put a bit more time in.

> As new shell modules can be introduced in different parts of the zephyr code base, it is quite difficult to keep track of them to add needed config options so that `SHELL_ALL` option works as expected.

@jukkar I agree and if someone has an idea how to improve this, please note it. My hope is, that the unified pattern and the documentation help to inform developers how they should implement their shell-module (with `default y if SHELL_ALL`).
